health/health_test.go

```diff
@@ -109,67 +139,114 @@ func TestRemoveAllWarnings(t *testing.T) {
 	if len(ht.Strings()) != 0 {
 		t.Fatalf("after RemoveAll, len(newTracker.Strings) = %d; want = 0", len(ht.Strings()))
 	}
+	if err := eventbustest.ExpectExactly(watcher,
+		compareChange(Change{WarnableChanged: true, Warnable: testWarnable}),
+		compareChange(Change{WarnableChanged: true, Warnable: testWarnable}),
+	); err != nil {
+		t.Fatalf("expected events, got %q", err)
+	}
 }
 
 // TestWatcher tests that a registered watcher function gets called with the correct
 // Warnable and non-nil/nil UnhealthyState upon setting a Warnable to unhealthy/healthy.
 func TestWatcher(t *testing.T) {
-	ht := Tracker{}
-	wantText := "Hello world"
-	becameUnhealthy := make(chan struct{})
-	becameHealthy := make(chan struct{})
-
-	watcherFunc := func(c Change) {
-		w := c.Warnable
-		us := c.UnhealthyState
-		if w != testWarnable {
-			t.Fatalf("watcherFunc was called, but with an unexpected Warnable: %v, want: %v", w, testWarnable)
-		}
+	var unregisterWatcher func()
+	tests := []struct {
+		name    string
+		preFunc func(t *Tracker, bus *eventbus.Bus, fn func(Change))
```

Comment:
I'm worried about the shared state between tests in `unregisterWatcher`. Could we maybe, say, have this pre-func accept the `*testing.T` value and, if it needs to, register a Cleanup on that `t` value to call the unregister function? Then the subtest will handle it, if necessary, and no shared state is needed from one to the next.